spec/models/jwt_api_entreprise_spec.rb

```diff
@@ -35,6 +35,21 @@
     end
   end
 
+  describe '.unexpired' do
+    subject { described_class.unexpired }
+
+    let!(:expired_jwt) { create_list(:jwt_api_entreprise, 2, :expired) }
+    let!(:unexpired_jwt) { create_list(:jwt_api_entreprise, 2) }
```

Comment:
Pour explicitement marquer l'opposition, nous pourrions ajouter un trait `unexpired`. Même si je suis certain que par défaut les tokens crées ne sont pas expirés :)

```suggestion
    let!(:unexpired_jwt) { create_list(:jwt_api_entreprise, 2, :unexpired) }
```

Reply:
je pense que ce trait n'a aucun cas d'usage, c'est le comportement par défaut sous entendu

Reply:
Le nom du let est déjà explicite imo. Je fais ce genre de chose pour clarifier au besoin, là ça me semble assez clair.

Après, j'ai pas plus d'avis.

Reply:
Je suis d'accord, c'est visiblement assez clair et on ne spécifierait ce trait `:unexpired` nul part ailleurs dans les specs donc je ne vois pas l'intérêt.

Reply:
Sauf que si typiquement on modifie la factory, qui est indépendante de ce test, ce test ne terstera plus ce qu'on attend.

Reply:
Cet argument est clairement entendable, mais à priori avec cette logique tu mets des traits sur l'intégralité des création de modèles ? (simple question voir la stratégie que t'adoptes derrière cet argument) Du coup à quel moment tu considères que tu peux utiliser ta factory sans trait ? Je pose cette question car ça m'est en effet arrivé de modifier une factory de base, ce qui m'a amené à nettoyer quelques tests à cause de ça.

Après faire ça:

```ruby
factory : jwt_api_entreprise do
  trait :unexpired

  # code...
end
```

ne coûte pas cher et peut en effet rendre ça plus clair.

Reply:
Voici la logique, dans l'ordre :

1. Ignorer les attributs inutiles
2. Renseigner aléatoirement les attributs requis
3. Regrouper par un trait le(s) attribut(s) sur une valeur particulière

En faisant comme ça, on a normalement moins de trait que d'attributs.

Note : à propos de la structure des specs, les traits sont pratiques car ils peuvent permettre de faire dériver chaque contexte, en n'ayant qu'une seule instance de modèle à déclarer.

Avec un exemple :

```ruby
# BAD

let(:var_a) { create_list(:foo, :trait_a) }
let(:var_b) { create_list(:foo, :trait_b) }

it 'returns value a' do
  expect(var_a).to be_value_a
end

it 'returns value b' do
  expect(var_b).to be_value_b
end

# GOOD

let(:var) { create_list(:foo, trait) }

context 'when a' do
  let(:trait) { :trait_a }

  it 'returns value a' do
    expect(var).to be_value_a
  end
end

context 'when b' do
  let(:trait) { :trait_b }

  it 'returns value b' do
    expect(var).to be_value_b
  end
end

# SUPER GOOD

let(:var) { create_list(:foo, trait) }

context 'when a' do
  let(:trait) { :trait_a }

  it 'returns value a' do
    expect(var).to be_value_a
  end

  it 'does not return value b' do
    expect(var).not_to be_value_b
  end
end

context 'when b' do
  let(:trait) { :trait_b }

  it 'returns value b' do
    expect(var).to be_value_b
  end

  it 'does not return value a' do
    expect(var).not_to be_value_a
  end
end
```

Reply:
> Sauf que si typiquement on modifie la factory, qui est indépendante de ce test, ce test ne terstera plus ce qu'on attend. 

J'entends clairement l'argument et je vois l'intérêt dans le cas de record avec des states en base un peu plus tricky. Ici c'est sur un unique champ et pas n'importe lequel : si demain on modifie la factory en question pour que les JWT soient par défaut expirés soit : 

1. ça ne fait aucun sens et *heureusement* que les tests vont casser 
2. la logique métier derrière les jetons a complètement changé et de toute façon faudra ré-écrire la suite de tests. 

Sinon, je préfère au contraire utiliser des factories par défaut plutôt que des factories setups pour des cas de tests précis et dont la création en base est complètement décorrélée du comportement de production, et je ne pense pas que ce soit une mauvaise pratique au contraire. Idéalement, on n'utiliserait `FactoryBot` uniquement pour des states complexes et bien précis, dans le cas de ce test il serait mieux de faire : 

```ruby 
let(:unexpired_token) { JwtApiEntreprise::Operation::Create(params: valid_params) } 
```

Ainsi, j'ai la certitude que mon record de test est iso avec ceux qui sont créés en environnement de production. C'est plus contraignant, il faut aussi savoir faire des compromis. Quoi qu'il en soit, c'est à mon avis une meilleure pratique plutôt que de setup des records à la main, exactement comme on souhaite qu'ils soient pour que le test soit vert, car ça introduit le risque de faux positifs.

Reply:
> si demain on modifie la factory en question pour que les JWT soient par défaut expirés, ça ne fait aucun sens et _heureusement_ que les tests vont casser

Les tests vont effectivement casser, sauf si il y a un bug dans l'application parce qu'on modifie aussi le code de l'application :trollface:

Sur ta proposition d'utiliser un constructeur à la place d'une factory, c'est intéressant. Fondamentalement, là aussi la dérivation par contexte est faisable. Nous pourrions avoir quelque chose comme ça :

```ruby
let(:var) { JwtApiEntreprise::Operation::Create(params: default_params.merge(foo: particular_value)) }

context 'when a' do
  let(:particular_value) { 'a' }

  it 'returns value a' do
    expect(var).to be_value_a
  end

  it 'does not return value b' do
    expect(var).not_to be_value_b
  end
end

context 'when b' do
  let(:particular_value) { 'b' }

  it 'returns value b' do
    expect(var).to be_value_b
  end

  it 'does not return value a' do
    expect(var).not_to be_value_a
  end
end
```